### PR TITLE
[rl] Exclude non-finite old-policy tokens from the global loss denominator

### PR DESCRIPTION
**Pull request by taking-lying-flat**
## Summary

- redefine the existing `num_global_valid_tokens` field to count response tokens with finite generator logprobs
- exclude response tokens with non-finite generator logprobs from the DAPO/GRPO loss and trainer-metric denominators
- keep `loss/generator_logprob_nan_frac` exact at batch-packing time without adding a second `TrainingBatch` field or trainer RPC argument

## Root cause

DAPO removed non-finite old-policy tokens from the loss mask but continued dividing by the pre-filter response-token count. This systematically reduced the policy loss and gradients when invalid generator logprobs were present.

## Design

The existing global valid-token count now uses `loss_mask & isfinite(generator_logprobs)`. The loss, importance-ratio metrics, and trainer entropy all use the same effective mask and denominator, so gradient accumulation matches a single large batch.

The response-level non-finite fraction still uses all response tokens as its denominator, but is computed once by the batcher. This preserves the metric's definition while avoiding a second token-count field, as requested in review.

## Test plan

- local targeted DAPO/GRPO numerical checks, including the all-non-finite edge case (5 passed)
- flake8, ufmt, pydoclint, codespell, AST, license, and whitespace pre-commit hooks
- targeted pyrefly check for the DAPO loss (0 errors)

### CODE REVIEW

File: torchtitan/experiments/rl/controller_metrics.py [hunk 1]
```diff
@@ -80,7 +80,7 @@ def combine_microbatch_metrics(
 
 
 def compute_perf_ratio_metrics(
-    *, num_global_valid_tokens: int, time_metrics: list[m.Metric]
+    *, num_global_response_tokens: int, time_metrics: list[m.Metric]
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/controller_metrics.py]: It seems this is the only place where non-finite count is used. Given this is not a critical quantity, can we just change the name / description of this variable to only include finite log-prob tokens as well?
- it will significantly reduce the change of this PR
- for vllm metrics, we anyway will do more fine-grained things like https://github.com/pytorch/torchtitan/pull/3958

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/experiments/rl/controller_metrics.py (hunk 1 above)
Comment (taking-lying-flat) [reply torchtitan/experiments/rl/controller_metrics.py]: lgtm. okk

File: torchtitan/experiments/rl/components/batcher.py [hunk 2]
```diff
@@ -155,11 +155,21 @@ def _pack_one_training_batch(self) -> TrainingBatch:
         # Next-fit all taken training_samples into rows.
         rows = self._assign_training_samples_to_rows(training_samples)
         packed_rows = [self._pack_training_sample_row(row) for row in rows]
+        num_global_response_tokens = sum(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py]: my point is that we don't need to introduce this field

File: torchtitan/experiments/rl/components/batcher.py [hunk 3]
```diff
@@ -155,11 +155,21 @@ def _pack_one_training_batch(self) -> TrainingBatch:
...
+            int(row["loss_mask"].sum().item()) for row in packed_rows
+        )
+        num_global_valid_tokens = sum(
```
Comment (tianyu-l) [inline torchtitan/experiments/rl/components/batcher.py:158]: but just adapt the definition of this field, like what this commit is doing now

Comment (taking-lying-flat) [reply torchtitan/experiments/rl/components/batcher.py:158]: Thanks for clarifying. Updated in fc7ec1ac1: `TrainingBatch` now keeps only the existing `num_global_valid_tokens` field, whose definition is `loss_mask & isfinite(generator_logprobs)`. I removed `num_global_response_tokens` and all additional trainer RPC/loss plumbing. Loss, ratio, entropy, and throughput now use the valid-token count.

To keep `loss/generator_logprob_nan_frac` mathematically correct without another field, the batcher computes that response-level metric locally before constructing `TrainingBatch`. Could you take another look?

**Review by tianyu-l — APPROVED** [review]
